File: Cpp/Source/LeaderBoard/source/Race.cpp

#include "../include/Race.h"
#include "../include/SelfDrivingCar.h"

#include <typeinfo>

const int Race::POINTS[] = { 25, 18, 15 };

Race::Race(std::string const& name, std::list<Driver> const& drivers) : m_name(name), m_results(drivers)
{
    for (Driver& driver : m_results)
    {
        std::string driverName = driver.getName();
        if (typeid(driver) == typeid(SelfDrivingCar))
        {
            driverName = "Self Driving Car - " + driver.getCountry() + " (" + (static_cast<SelfDrivingCar&>(driver)).getAlgorithmVersion() + ")";
        }
        m_driverNames[driver] = driverName;
    }
}

int Race::position(Driver const& driver)
{
    int count = 0;
    for (Driver& d : m_results)
    {
        if ( d == driver ) 
        {
            break;
        }
        ++count;
    }
    return count;
}

int Race::getPoints(Driver driver)
{
    return POINTS[position(driver)];
}

std::list<Driver> Race::getResults()
{
    return m_results;
}

std::string Race::getDriverName(Driver driver)
{
    return m_driverNames[driver];
}

std::string Race::toString()
{
    return m_name;
}

File: php/src/Leaderboard/Race.php
<?php

declare(strict_types=1);

namespace RacingCar\Leaderboard;

class Race
{
    /**
     * @var int[]
     */
    private static array $points = [25, 18, 15];

    /**
     * @var array<string, string|null>
     */
    private array $driverNames;

    /**
     * @param Driver[]  $drivers
     */
    public function __construct(
        private string $name,
        private array $drivers
    ) {
        $this->driverNames = [];

        foreach ($drivers as $driver) {
            $name = $driver->name;
            if ($driver instanceof SelfDrivingCar) {
                $name = "Self Driving Car - {$driver->country} ({$driver->algorithmVersion})";
            }
            $this->driverNames[(string) $driver] = $name;
        }
    }

    public function getPosition(Driver $driver): int
    {
        return array_search($driver, $this->drivers, true);
    }

    public function getPoints(Driver $driver): int
    {
        return self::$points[$this->getPosition($driver)];
    }

    /**
     * @return Driver[]
     */
    public function getResults(): array
    {
        return $this->drivers;
    }

    public function getDriverName(Driver $driver): string
    {
        return $this->driverNames[(string) $driver];
    }
}
